staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go

```diff
@@ -83,6 +84,16 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
 		return nil, err
 	}
 
+	// roundtrip aggregator spec because it is the only one which does not go through the delegate HTTP handler.
```

Comment:
this is hard to follow... why is the spec returned from BuildOpenAPISpec unsuitable? can we fix the output of that function?

Reply:
The aggregator spec had non-nil, empty `Scheme.Properties`. After roundtrip these turned into `nil`s for all other delegated apiserver specs (those we download via HTTP handlers, yes crazy, including JSON unmarshaling, at least with etags (I hope); different topic). The consequence is that the merge algorithm thinks the definitions are different, which leads to renames and duplication of all meta/v1 structs.

Reply:
I added a [quick unit test](https://gist.github.com/liggitt/37c3d3b798bf0b392f389dd74458890b#file-openapi_test-go) to see what the diff between in-memory and round-tripped schemas were, and the [results](https://gist.github.com/liggitt/37c3d3b798bf0b392f389dd74458890b#file-openapi_test_results-txt) were scary. In addition to the nil vs `[]`/`{}` example, it looks like additionalProperties is opposite in memory and once round tripped.

Rather than ad-hoc fixup here, can we adjust the generator to produce things that roundtrip without changes?

Reply:
what does "opposite" mean here?

> can we adjust the generator to produce things that roundtrip without changes?

I don't think is possible. Vendor extensions are just `interface{}`s. There is no scheme like in kube which could translate them back to native structs, you will just get some JSON blob instead.

Reply:
> what does "opposite" mean here?

in-memory, additionalProperties is `false`, which gets omitted when serialized, then defaults to `true` when deserialized